### ISSUE TITLE
replication error: failed to create chunk (Chunk already exists)

### ISSUE THREAD

**Issue by guestisp** [lizardfs/lizardfs#604]
Trying to replace an hdd with another one on the same chunkserver and getting this:

`replication error: failed to create chunk (Chunk already exists)`

I've marked for removal `/export/sda` and added `/export/sdb`.
I think lizard is refusing to move from sda to sdb because are both on the same chunkserver.

This is not good, because i'm no trying to create multiple chunks on the same chunkserver, but i'm trying to move chunks from a disk to another one. In this case, chunks should be allowed to coexists on the same server.

**Comment by 4Dolio** [comment]
Ya.. as of 2.6.0 anyway, the mark for removal does not do what you might expect.. it only seems to prevent additional writes to that device.  It does not 'evacuate' the data to a new local nor remote disk.

You should have more than one copy of all chunks. So you can comment out the old disk completely, then replication will recover the lost chunks.

Or. You can manually copy your chunks from your sda to your sdb.

**Comment by guestisp** [comment]
3.10 is moving chunks away, but still try to delete them from the "old" disk.
If disk is set RO by the underlaying FS (or by sysadmin), it logs tons of messages.

But other than this, seems to be unable to move chunks to another disk to the same server. 
I think this could be allowed. The constraint is good for normal goal management, but in case a disk is marked for removal and another one is available on the same server, that should be used.

**Comment by 4Dolio** [comment]
Unfortunetly i do not think lfs ever replicates or actively rebalances between disks on a single CS. 

Never has, not that it would not be useful, probably not a trivial thing to add however.

On a single CS I have gone so far as to write fractional rebalance scripts.  I can use it to roughly redistribute, based on chunk file size, say from one source disk 1/5 of chunks to each of 5 new target disks.

Or. Just pull the disk and let replication do it's thing...

**Comment by 4Dolio** [comment]
If you disk is RO then you need to fix that, it is not the fault of LFS.  It should mark the disk as failed with the marking it for removal mechanics.  You must then take appropriate actions to completely solve the problem.

**Comment by 4Dolio** [comment]
LizardFS does it's job very well.  But it can never account for all possible failure modes and be expected to know what the appropriate action to take is.  You have to leverage it for the tool that it is and build upon it and with it given the limitless variations of possible environments in which you might use it.  It can take some time to get up to speed with what LFS does and does not do.  Understand it's limitations and address them given your resources and goals.

**Comment by guestisp** [comment]
If I manually set a disk as RO, is not a fault and there is nothing to fix

It's the same as replace a disk in a raid without removing the older one but adding a new one, wait for rebuild and then remove the older one.
In this way you'll never loose redundancy, because you'll never run with a degraded array. At any time, during a replace, you'll run with increased redundancy

If you remove disk for replacement, you'll run in degraded mode for the whole operation

What I would like to do is the same but without replicate over network. If I have a working disk that I would like to replace with a new one (newer? Bigger? Faster? It doesn't matter) I would like to mark the old one for removar, wait for a local sync and then remove the old disk

Local sync is faster and doesn't involve any network activity that could lead to a performance drop

Even better would be to mark a disk for replacement, where , for the whole operation, goal increase of 1 from the defined one and only after a successful replacement the old disk is decommissioned. Allowing, for this operation, the use of the same chunkserver.
As lizard store everything in small chunks, every moved chunks should result in deletion from the old disk (except in case of RO filesystem)

I think this is the way how actually lizard works except for the chunkserver constraint

**Comment by 4Dolio** [comment]
I hear ya.  But there is no local-local replication mechanics as far as I know.  So, the solution is:
`-/export/sda # mark for removal`to prevent any new writes
cp -a /export/sda/* /export/sdb # Do a fast local copy, or rsync, or whatever you like
When it finishes:
`#/export/sda # remove`
`/export/sdb # add`
Restart/reload the CS

**Comment by guestisp** [comment]
Mark for removal doesn't prevent only new write, it start to remove files by replicating them somewhere else.
So, it's impossible to mark a disk for removal without setting it as RO and getting flooded with tons of error messages in log file.

**Comment by guestisp** [comment]
You can't do what are you saying. If you set the mountpoint as RO, you'll be flooded by error messages.
If you don't set that as RO, Lizard starts to remove files thus your copy won't be consistent.

**Comment by 4Dolio** [comment]
What you saw is normal. and expected.

I did not say read only in my procedure.  You could do that to preserve a rouge/empty metadata master from trying to delete chunks with read only or some other file retention mechanic.  That might be something to plan for anyway. You never know, never trust any single tool blindly or completely?

As you manually copy from the marked disk it will slowly write newly written existing chunks to other disks and remove the marked chunk copy.  No new writes will occur on the marked but it should be able to delete chunks as needed.  It may or may not do some slaw lazy replication but it does not change how aggressively it prioritizes that. If it did replicate it would likely be due to some other rebalancing schedule or other triggers.

If you have extra chunks that were the deleted, or you indexed a chunk file which no longer exists once you get around to trying to copy it.  If is no mater, extra chunks get purged on scan, and missing ones are already elsewhere and safe.

A new feature similar to 'marked' could be 'eject', it could be an isolated CS action internal to a discrete CS service to quickly move chunk away from the 'ejected' disk. An additional hdd.cfg line prefix of + perhaps indication you want to be +1 disk in your hand asap, would still take some time but be reflected in the cgi and cli tools.

Another 'abandon' status could be auto chosen if logic can be predicted, or you could issue to the master such a status change for a disk.  This would cause the rapid abandonment of the disk by replicating away as if the impacted chunks did not exist and caused undergoals. The master would still consider their existence and direct CS to CS replication to utilize them if needed.  The true disk chunk counts should still reflected in the cgi and cli, but perhaps the goals would reflect as if the chunks were already missing.

**Comment by 4Dolio** [comment]
So.. you can easily manually 'eject' or 'abandon' a disk if you know what to expect of how to prepare.

**Comment by guestisp** [comment]
@4Dolio currently, it's impossible to replace a disk by marking it for removal.
I've tried right now. I only have 3 chunkservers in this test cluster and I've set goal=3.

I've added a new disk on any chunkserver but this new disk won't be used at all, as will result in multiple chunks stored on the same chunkserver.

Marking an hdd for removal should move chunks from disk to be removed to another one, but as I don't have any other chunkserver available, it does nothing.
What I thought is that marking a disk for removal would allow chunk movement even between the same chunkserver, that's because doesn't produce any loss of redundancy. chunks was already on that chunkserver, i'm only asking to move from hddA to hddB

I can't manually sync these 2 disks, because lizard starts to delete chunks (if a chunkserver is available) thus, a simple `cp -a /hddA /hddB` will result in big inconsistencies.

Any solution without bringing down the whole node ?
Let's assume a fourth chunkserver available. Marking a disk for removal will start a network replication and chunks deletion from the local disk. This add an unnecessary load on network and makes impossible to manually sync both disks with `rsync` or `cp` due to the ongoing chunks removal.

You have to bring down the chunkserver, triggering again a network replication.

So, there is no way...

**Comment by Blackpaw** [comment]
> I can't manually sync these 2 disks, because lizard starts to delete chunks (if a chunkserver is available) thus, a simple cp -a /hddA /hddB will result in big inconsistencies.

What sort of inconsistencies? a temporary doubling up of chunks won't matter.

**Comment by guestisp** [comment]
> What sort of inconsistencies? a temporary doubling up of chunks won't matter. And using rsync rather than cp will handle any issues with chunks being locked or deleted.

marking a disk for removal starts to delete chunks from the marked disk.
You can't start to copy (with rsync or cp) if Lizard is removing chunks, as you'll end with a partial copy and not a full copy.

> Marked for removal does not do what you think it does. It does not induce any replication.
Again, YES. Marking a disk for removal start to replicate away chunks, that's the main purposes for marking a disk.

It's also the same in MooseFS:
https://moosefs.com/documentation/faq.html#6
And even in Lizard:
https://docs.lizardfs.com/adminguide/basic_configuration.html#configuring-your-chunk-servers
```
To remove a directory from being used by LizardFS, just add a * to the beginning of the line in mfshdd.cfg:

# use directory '/mnt/hd2', but replicate all data from it:
*/mnt/hd2
LizardFS will replicate all the data from it somewhere else. Once you see in the web interface that all data has been safely copied away, you can update the file and remove the line and than remove the device associated with it from your chunkserver.
```

So, @4Dolio, please stop saying that marking a disk for removal doesn't trigger any replication.
Mark for removal TRIGGER A REPLICATION, it's supposed to do that.

And if you are replicating aways chunks, you can't obviously, run an rsync process to keep things synced, because they won't be synced. You'll end with a partial copy. Rsync is able to copy only chunks not already removed by lizardfs.

**Comment by Blackpaw** [comment]
Ah, I see, sorry, misunderstood the process you were describing.

Sounds like your best option currently is to:
- stop the chunkserver
- rsync the data from HDD1 to HDD2
- update mfshdd.cfg to point to HDD2
- restart the chunkserver

Unless, is the existing HDD in a zfs pool? if so you could 
- add the 2nd hdd as a mirror, 
- wait for the resilver to finish
- detach disk1

That would involve no down time.

**Comment by guestisp** [comment]
I don't want to stop a whole chunkserver because:

1. This will trigger a new network replication
2. This will bring down all chunks and not only the ones stored on disk that need to be replaced

In other words, bringing down the chunkserver is even worse as it start to replicate the whole server and not only a single disk

If you have to transfer a 4/8tb disk, you'll need some hours that will trigger a replication even with standard lizard timeouts

**Comment by 4Dolio** [comment]
The plan i described does not include stopping the CS at any point.  Only one soft HUP, reload.

**Comment by 4Dolio** [comment]
My mistake. So it does replicate. Forgive me.

My plan still works.

The replication is slow.. and you can tune it to be slower or wait 24 hours after a CS registers.  Rep delay settings.

Rsync will never create a partial copy.. nor would cp i do not think.. it will fail to find some source chunk files, but you can ignore those...

Then just reload to remove old and add new disk after manually copying disk to disk..  It works. Done it. Done crazier versions of it.. done ot between 10 pairs of nearly full 4TB disks, in parallel, overnight.

**Comment by 4Dolio** [comment]
Sorry about insisting it does not replicate... but it is still not appropriate to make the disk read only.  And the sloppiness of the replication+rsync failing to find removed files will not matter at all in the end and can be completely ignored.  Copy/rsync would never result in a partial copy.. the operation will either not find the file inode if deleted.. or it will find the file and it will be locked and the whole file transfered will complete, even if the inode is unlinked by the removal by lfs of the chunk file.

**Comment by 4Dolio** [comment]
... maybe the inode thing depends on your filesystem?  Anyway, I still dont think rsync or even cp would write a corrupt file.. you can test that on a big file that takes time to copy.. copy it and delete it before it finishes copying, what do you end up with?

**Comment by guestisp** [comment]
I'm not talking about corrupted files but if you have to transfer 50.000 files and one process is deleting them during your copy, you'll end with less than 50.000 files copied because some of them was deleted by the other process before your copy

Is not that hard to understand.....
You can't do a `rm -rf * ; rsync ...` hoping that rsync will catch all files before rm.....

And, again, I DON'T WANT TO PUT AN EXTRA LOAD ON NETWORK, so I can't use lizard replication...

Probably, the best solution would be to fix a bug (as reported above by me and by @onlyjob) and maybe implement a local replication, again reported by @onlyjob instead of trying with some bad and ugly workarounds that won't solve the issue at all but put additional load on network and requires at least one spare chunkserver that I don't have 

So, with 3 chunkservers only and goal=3 there is no way to change a disk without going undergoal for many hours with lizard (remove the old HDD from mfshdd.cfg, this makes you undergoal, start rsync, add the new HDD to mfshdd.cfg and now you'll be back with proper goal) but this doesn't involve any network transfer

Even if you just physically remove a disk, you'll go undergoal for the whole sync operation and lizard is forced to sync via network (due to the missing local disk with all chunks that needs to be replicated) and could last for tens of hours depending on disk size

The ability to locally transfer a disk marked for removal would fix all of these issue at once

**Comment by 4Dolio** [comment]
Looking at #364 It appears that I agreed and offered the same local copy work around.

I do agree in #348 that lfs should more gracefully handle read only filesystems, but I do not feel it is a fatal problem.

In either case.. I an not a contributor or developer but a mere user of LizardFS.
Those are both from 2015, they are not likely to be resolved in the next weeks/months.
It is what it is.  I have encountered these hurdles myself and am attempting to help.  I am only human and I do make mistakes.  But I have kicked the snot out of many aspects of LFS and think I have a pretty good feel for how it will react to things you are asking about. 

The way I see it.  You have three options:

1. Submit a pull request to solve those shortcomings.

2. Tune down your replication rate so it does not saturate your disks or network or whatever it is you are afraid will happen if you just pull the bad disk and add the new one.  Let LFS replicate at a reasonable rate that does not cause problems.  You could even preemptively raise your goals in anticipation of losing some chunks if you fear being under goal and less redundant.  So let LizardFS do what it currently does and just waste the bits on the wire, They cost you nothing but time.

3. QUIT YELLING AT ME.  Try to go over the process I described and tell me what won't work again.
A. It DOES NOT MATTER that replication is happening away from the marked for removal disk.
B. Keep the disk writable, or make it read only and ignore the errors. It won't matter either way.
C. It Does NOT Matter that you will fail to copy some files because replication removed them.
D. It Does NOT Matter that some files will be duplicated from being both copied AND replicated.
E. You can minimize replication if you fear it will load your network. MaxRead/WriteRep=1
... Or replication delay=weeks
... If replication is breaking you network, then you need to address your replication tuning rates anyway.
... If you have rep read/write =1, and 5 minute loops, it should do 288 chunks per day.
... If you have rep read/write high enough that it causes problems then it is tuned to high.

Once the copy has finished:  Pull the old disk and online the new disk.
* The chunks you filed to copy are already replicated safely elsewhere.
* The chunks you have duplicates of on the new disk will be purged during the reload.
** This will slow down the disk scan process depending on how many dups there are.
** Since I was wrong about the marked for removal, I suggest you discourage replication.

Note: This would even work if you do not mark the old disk for removal.
* Assuming you have minimum of goal=2
* 90%-99% of the chunks will be unchanged. Depending on how active your data is.
* After 24 hours to push 4TB between local disks, some chunks will of course be incorrect.
* LizardFS will purge duplicates when scanning the new disk. It should also identify old stale chunks.
* LizardFS will then quickly replicate the missing 10%-1% of chunks which were changed in a few minutes.

So. Keep on waiting for LizardFS developers to address those two year old issues.  Fix them yourself.  Let LizardFS just take care of it and tune your cluster to not self implode.  Try to hack the crap out of shit until you know what works and what does not work, cheat on this last one by listening to me.

**Comment by 4Dolio** [comment]
If I had not done what you are trying to do many times myself, I would not be trying to help here.  Sorry I never use the marked for removal feature or that my clusters are tuned such that their replication never kicked in due to marked due to singular labels or long replication delays.  When something goes wrong I have enough redundancy to stay online long enough to look closer into the problem and resolve it, which is fairly rare.  I fix the problem and LFS just hums along does it's job and rolls with it.

I have been using MooseFS and LizardFS since 2011 and have about a dozen or so production LFS clusters on three continents.  Metadata sizes ranging from 100KB to about 50GB.  Approaching a billion objects. 500+GB of RAM for masters.  Maybe about 100 spindles and half as many solid disks.  Hosting shares, websites, revision control, iSCSI, Multiple Bare Metal and Virtual Instances of all roles in the clusters.  At home I manage to store millions of files in clusters run on mere 512MB a Raspberry Pi has to offer and whatever junk disk I happen to have.  I once ran a Lizard Cluster on a pair 1/8" audio jack serial linked TI-82's (just kidding).

**Comment by Blackpaw** [comment]
>At home I manage to store millions of files in clusters run on mere 512MB a Raspberry Pi has to offer and whatever junk disk I happen to have.

What sort of data? media storage? How many Pi's do you use with that. Are you connecting your disks via the Pi usb bus?

**Comment by 4Dolio** [comment]
20 years of photos, mp3 collection, home VHS movie files. past project data. installers. backups of phones. backups of servers. other garbage.  All the junk you would find on a data horders drives, about 4TB worth.  Next I'll try to automate part of the deduplication of stuff that has been backed up more than once. As a last resort I'll throw away my dos6.2 and win3.1.1 floppy backups. ;)  Currently running 2 Raspberries, a Bananna, and an ODroid2, and a few junk laptops, still need to figure out how to integrate a few old Palm Pre WebOS phones.. It isn't very pretty and things change roles all the time, haven't yet tried the corosync HA from work yet.  Disks are on USB, yes, It only has one client so it is fine.  The trick to make it usable on a Pi is nesting filesystems.  Loopbacks or iscsi or whatever container method you want to use, keeps the metadata size tiny and you can still leverage things like zfs compression.

I really appreciate when projects KISS rather than trying to do every possible thing they 'could' do.  Makes such tools more versatile and allows them to be leveraged in all sorts of unexpected and unpredictable ways.  I'm not sure I'm ready for LizardFS to be all grown up and version 6+ with built in everything. sigh, I'm sure it will be fine.
